### DISCUSSION TITLE
Utilities for surfacing styles

### DISCUSSION THREAD

**Discussion by KingSora** [remix-run/remix#4191] [Proposals]
I personally think that remix is doing a fantastic job in "not hiding how things work" and I'm a fan of that, but at the same time I think remix is making it hard / very verbose to have scoped styles. So I've been  digging into some kind of automation of collecting the stylesheets which should be surfaced and I think I found a elegant solution to it. With this post I want to gather feedback from you guys and wanna know whether you think this could be a thing which is worth to have "native" support in remix:

## **1. Whats the problem:**
In an modern web app I want to have component scoped stylesheets and I don't want to take care of re-exporting them correctly all the time.

### **1.1. How is it done now?**
Like described here: https://remix.run/docs/en/v1/guides/styling#surfacing-styles

### **1.2. Why is the current solution not optimal?**
Developers have to take extra care with stylesheets handling and the code ends up very verbose.

## **2. Proposed solution:**
What I'm proposing is not to replace or get rid of any of remix concepts, I just want to have "helpers" which makes it more convenient to use remix concepts such as the links export.
So what if we could analyze each route and detect the used stylesheets automatically. This detected styles are gathered in a virtual module, which exports them as an array. This virtual module would look something like this:
```js
import { default as s0 } from "../assets/button.css";
import { default as s1 } from "../assets/list.css";
import { default as s2 } from "../assets/checkbox.css";
export default [s0, s1, s2];
```
Note that the array has the correct order of styles according to the import order of the styles. Now that we have this information it only needs to be exposed. This can be done however you think is best, but I'll propose it as a simple import. Something like `import styles from './index.<suffix>.js';`. This file is not there obviously, its just a virtual one which is resolved by `esbuild`.

Now instead of surfacing each style component by component we would end up with an correctly sorted array of stylesheet paths, which the consumer can use. Its by the way completely up to the consumer whether to use this strategy or not, this is 100% additional and is not breaking anything.

The consumption would look something like this:
```js
import styles from './<routename>.<extension>';

export const links() => [
  ...styles.map((style) => ({ rel: "stylesheet", href: style })),
]
```
Since the consumer has still full control over everything he could decide to just use a fraction of the array, or to use any `media` options etc. This is really just an helper for getting the urls, its not doing anything with them and if the helper is unused it gets treeshaken.

### **2.1. Implementation:**
The implementation is done with `esbuild` and its `css` loader. Esbuild is able to generate a css file for each entry point (https://esbuild.github.io/content-types/#css-from-js), this css file doesn't have to be written to disk (`write: false`) and can be analyzed in the metafile object (https://esbuild.github.io/api/#metafile). In this metafile object you'll find all the information needed (all the used stylesheets in the correct order). After this information is gathered we just need to write the virtual-module and we are finished.

### **2.2. Implications:**
With this utilities it would be very easy to implement other stylesheet based css systems like e.g. `css-modules` because they only need to be imported in order to show up in the virtual modules array export. 

The resulting behavior per route would be that each route consists not of one but possibly more stylesheets depending on how long the array is. The benefit of that is that **shared stylesheets can be cached** and other routes with similar stylesheets doesn't have to download everything. 

Other benefit are **CSS Order** and **No unused CSS**, because each route has its own array, the order is correct _per route_ and in comparison to one big bundled css file where half of the styles could be unused we have certainty that all of the styles in the array are used in this route.

### **2.3. Possible Improvements:**
This section only describes what could be added on top of this system. Its just me thinking about possible future features, this is not part of the proposal and only shows what could be possible:
Since the virtual module is exporting an array of stylesheet paths, it could be decided to add more information to the array if needed. There could also be a field in the remix config where someone can adjust how the array is generated and what data are in it.

## **3.0. Negatives**
Obviously this system brings overhead, but since remix needs to process its stylesheets anyways to fix bugs (https://github.com/remix-run/remix/pull/4130) this could be implemented right into this step. Additionally the implementation logic is only "run esbuild and use its output", which is not a very complex behavior in comparison what it could be with other things.

**Comment by cliffordfajardo** [comment]
@chaance from the Remix Team was prototyping CSS modules support in https://github.com/remix-run/remix/pull/2489 he probably could weigh on this esbuild approach you're proposing

**Reply by KingSora** [reply]
@cliffordfajardo would also be ready to do a POC for demonstration

**Comment by chaance** [comment]
Really solid proposal, IMO. Thanks for the explanation as well!

This is actually somewhat similar to our plan for CSS Modules, with the obvious difference being the modules part.

We had planned on starting with a virtual module that compiles everything into a single module that could be imported in the `root` instead of splitting them by routes. That's not because there isn't value in your approach; ideally the app would only care about the styles needed on a given route. The downside here is really DX. This forces the developer to remember to import the stylesheet on every single route, which is both clunky and error prone, and we haven't come up with a good workaround to simplify things just yet.

**Reply by KingSora** [reply]
@chaance as a side note, I created the proposed system for my own purposes which are `css-modules` in my project, so I spend a bit of time evaluating a few solutions. I also tried a the approach you explained, but found some issues with the **css-order**:

I understood every route in Remix as a separate entry point. That means If I decide to import lets say `Button` and `List` in that order, the css-modules order is exactly like that for this route. If I'm in an different route and decide to import `List` and `Button` in this reverse order, the css-modules order should also adapt to that. Obviously its also possible to create a virtual module which imports every possible route like `import 'routes/**/*';` and thus can emit a single css file for all of them. Here the css-modules order is determined by the virtual module import order and as a consumer you don't have control over this process, which could lead to unexpected behavior.

Regarding:
> The downside here is really DX. This forces the developer to remember to import the stylesheet on every single route, which is both clunky and error prone, and we haven't come up with a good workaround to simplify things just yet

I think the solution for this should be straight forward and simple, and the simplest one (for me) is to have an reserved export name in each route just for css modules. This export is just for **internal use only** and shouldn't be documented publicly. Lets say the export has the name `export const __remixInternalCssModules = /* ... */` then you could build a `esbuild` plugin which adds this export automatically to every route. Then remix just needs to treat that export similarly to the `links` export and you would have an automated import for css modules in each route.

That said, If possible I would like to have the possibility to choose how css-modules work in Remix (one large vs many small css files) since all other frameworks like `Next` etc. have frequent bug reports regarding css order with css modules.

**Reply by kiliman** [reply]
I thought the whole point of CSS modules was that all names were scoped locally to prevent specificity issues. Wouldn't that mean that CSS source order should be a non-issue?

**Reply by sergiodxa** [reply]
Yeah the order shouldn't be an issue when using CSS Modules, that's part of the point of using them, you should be able to load the CSS in any way you want.

**Reply by KingSora** [reply]
@kiliman @sergiodxa I believe you're mixing up things now. Its correct that css modules will scope your selectors, but this has nothing to do with specificity.
Consider the following example:

```js
import stylesA from 'stylesA.module.css';
import stylesB from 'stylesB.module.css';

<div className={`${stylesA.container} ${stylesB.container}`} />
```
The `.container` style from `styleA` is `background: white` and from `styleB` its `background: black`.
If the css order is irrelevant, which background color would the div have, `white` or `black`?

The answer is: it depends on the import order. The stylesheet which is imported later is located further down in the output css file, which means it has a higher specificity.
I've created an example here: https://codesandbox.io/s/react-css-modules-forked-v6m6hz?file=/src/index.js

**Reply by kiliman** [reply]
I understand that it's possible, but why would you want to apply styles from 2 different containers? I thought the whole point was to scope CSS per component.

That's why I use Tailwind CSS. At least when I look at a `<div>` I know exactly which styles are applied and don't have to worry about esoteric rules.

**Reply by KingSora** [reply]
@kiliman this is just an example.. in reality this would happen like this:

```js
import Button from 'components/Button';
import styles from 'styles.module.css';

const MyLayout = () => {
  return <>
    {/* ..complex layout.. */}
    <Button className={styles.coolBtn}>
      cool button
    </Button>
  </>
}
```

Where the `Button` component has internal default styles which are overwritten by the passed styles. (But only if the imported styles are below the imported Button component)

**Reply by kiliman** [reply]
I understand. I think that just moves the problem somewhere else. I typically don't let my custom components accept the `className` prop. Instead, I expose various style variants: like size, color, etc. Then internally I map those to the appropriate TW classes.

Anyway, since I don't use CSS modules, I have no dog in this fight.

**Reply by KingSora** [reply]
@kiliman the `className` prop becomes important if you build general purpose components for your project or if you use an components library. 

All I wanna demonstrate here is how css modules work and what the dangers are.. So that the decisions which are made will improve the product
